### ISSUE TITLE
Should we keep a top-level list of "allowed" packages

### ISSUE THREAD

**Issue by dmose** [dmose/mc-nodejs-docs#15]
@loganfsmyth writes:

> Are we defining our own structure, or is this basically just committing `package.lock` + `node_modules` folder? For example, ehen vendoring a package, are its subdependencies flattened and also loadable directly, or is it only the specifically vendored top-level package that is loadable? If things are flattened, how do we handle multiple modules with differing subdependencies? If things aren't flattened, how do we handle duplicates?

We're just committing package.json, package-lock.json and node_modules.

> If things are flattened for deduplication, do we want to do anything to maintain a top-level list of packages we allow that projects actually use? For instance, one common accident to make in Node is to start using a node_modules dependency even though it isn't in your `package.json` (because it happened to be in `node_modules` due to being flattened out from inside some other dependency). Especially as the number of vendored packages and number of subdependencies grows, I can see it being very difficult to figure out what things are actually being used directly throughout the codebase vs as subdependencies, which could make it difficult to be confident that things can safely be removed from the set of vendored packages.

Isn't that list of things just the list of dependencies and devDependencies in package.json?

CC @ncalexan

**Comment by loganfsmyth** [comment]
> Isn't that list of things just the list of dependencies and devDependencies in package.json?

Yep! I should have clarified more that for this part I'm wondering  access restrictions. Should we make it so our commonjs loader only allows files outside the censored folder to access modules that are declared in the `package.json`?

Or an ESLint rule to error in the case of vendor imports to undeclared dependencies?

**Comment by dmose** [comment]
The ESlint rule sounds like a fine way to do this.  We'll need to talk to @Standard8 to make sure there aren't parts of the tree that are intentionally .eslintignored that we will miss.
